### PR TITLE
[https://nvbugs/5123103][fix] Fix torch compile for DeepSeekV3

### PR DESCRIPTION
**Pull request by liji-nv**
# PR title

Please write the PR title by following template:

[JIRA ticket link/nvbug link/github issue link][fix/feat/doc/infra/...] \<summary of this PR\>

For example, assume I have a PR hope to support a new feature about cache manager of Jira TRTLLM-1000 ticket, it would be like

[TRTLLM-1000][feat] Support a new feature about cache manager

## Description

Please explain the issue and the solution in short.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 1]
```diff
@@ -293,11 +295,12 @@ def __init__(
 
         try:
             if pytorch_backend_config.torch_compile_enabled:
+                torch._dynamo.config.recompile_limit = 16
```
Comment (hlu1) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py:298]: Why is this necessary?

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 2]
```diff
@@ -523,10 +526,13 @@ def _create_extra_inputs(bs, num_tokens_per_request):
                 if self.batch_size < 2:
                     warmup_batch_size = [1]
                 for bs in warmup_batch_size:
+                    self.spec_config.max_draft_tokens if self.spec_config is not None else 0
```
Comment (hlu1) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py:529]: What is this line doing? Is it dead code?

File: tensorrt_llm/_torch/model_config.py [hunk 3]
```diff
@@ -32,6 +32,10 @@ class ModelConfig(Generic[TConfig]):
     attn_backend: str = 'TRTLLM'
     moe_backend: str = 'CUTLASS'  # options can be CUTLASS, TRTLLM
 
+    mla_layers: Dict[int, object] = field(default_factory=dict,
+                                          repr=False,
+                                          init=False)
```
Comment (hlu1) [inline tensorrt_llm/_torch/model_config.py:37]: This config is too Deepseek specific and should not be part of generic `ModelConfig` .

File: tensorrt_llm/_torch/modules/attention.py [hunk 4]
```diff
@@ -208,6 +211,62 @@ def forward(
...
+    mla_layers = extra_attrs.get("mla_layers", None)
+    assert mla_layers is not None, "MLA layers is not registered"
+    mla_layer_ref = mla_layers.get(layer_idx, None)
+    assert mla_layer_ref is not None, f"Cannot find mla layer for layer {layer_idx}"
+    mla_layer = mla_layer_ref()
+    assert issubclass(type(mla_layer), MLA) or isinstance(
+        mla_layer,
+        MLA), "MLA layer must be a subclass of MLA or an instance of MLA"
+
+    return mla_layer.forward_impl_fake(hidden_states)
```
Comment (hlu1) [inline tensorrt_llm/_torch/modules/attention.py:267]: Duplicated code. Please deduplicate.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 5]
```diff
@@ -523,10 +526,13 @@ def _create_extra_inputs(bs, num_tokens_per_request):
...
                     for num_tokens_per_request in [
                             1,
-                            min(self.max_num_tokens // max(bs, 1),
-                                kv_cache_manager.max_seq_len - 1)
+                            min(
+                                self.max_num_tokens // max(bs, 1),
+                                self.max_seq_len - 1 -
+                                kv_cache_manager.num_extra_kv_tokens)
```
Comment (hlu1) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py:535]: Can you explain the math used here?

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 6]
```diff
@@ -547,8 +553,9 @@ def _create_extra_inputs(bs, num_tokens_per_request):
 
         if self.pytorch_backend_config.autotuner_enabled:
             with no_cuda_graph(), autotune():
-                num_tokens_per_request = min(self.max_seq_len - 1,
-                                             self.max_num_tokens)
+                num_tokens_per_request = min(
+                    self.max_seq_len - 1 - kv_cache_manager.num_extra_kv_tokens,
+                    self.max_num_tokens)
```
Comment (hlu1) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py:558]: Why is the math here different from line 532-535?

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 7]
```diff
@@ -1853,6 +1860,11 @@ def forward(self,
             return outputs
 
     def model_forward(self, **kwargs):
+        attrs = get_model_extra_attrs()
+        assert attrs is not None, "Model extra attrs is not set"
+        attrs["attention_metadata"] = weakref.ref(kwargs['attn_metadata'])
+        attrs["mla_layers"] = self.model.model_config.mla_layers
```
Comment (hlu1) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py:1866]: Can we think of a better way to do this? This is too hacky.

File: tensorrt_llm/_torch/model_config.py (hunk 3 above)
Comment (liji-nv) [reply tensorrt_llm/_torch/model_config.py]: It now stores extra_attrs so any layer can register useful information into config.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 5 above)
Comment (liji-nv) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: Factor out the code. The logic is similar to cuda_graph warm up.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 6 above)
Comment (liji-nv) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: torch_compile will compile different batch size while autotune only profile bs1.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 7 above)
Comment (liji-nv) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: The mla is too complicated to put everything to the custom op interface. This solution here is similar to how vLLM handle attention.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 1 above)
Comment (liji-nv) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:302]: Currently the recompile times is more than 8 for complicated network like DeepSeek.

File: tensorrt_llm/_torch/custom_ops/cpp_custom_ops.py [hunk 8]
```diff
@@ -361,3 +361,21 @@ def _(
...
+                                    (batch_size, ), dtype=torch.int32)
+
+    @torch.library.register_fake("trtllm::fp8_quantize_1x128")
+    def _(input: torch.Tensor):
+        pad_m = fp4_utils.pad_up(input.shape[0], 4)
+        sz = fp4_utils.pad_up(pad_m *
+                              ((input.shape[1] + 127) // 128) * 4, 128) // 4
+        return torch.empty_like(input,
+                                dtype=torch.float8_e4m3fn), input.new_empty(
+                                    (sz, ), dtype=torch.float)
```
Comment (hlu1) [inline tensorrt_llm/_torch/custom_ops/cpp_custom_ops.py:381]: Can you put this in a thop and call it here? You can call the same function here https://github.com/NVIDIA/TensorRT-LLM/blob/main/cpp/tensorrt_llm/thop/fp8Quantize.cpp#L58-L59

The shape computation is pretty complicated and that prevents it from going out of sync

File: tests/integration/test_lists/test-db/l0_h100.yml [hunk 9]
```diff
@@ -32,16 +32,21 @@ l0_h100:
...
-  - accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_fp8_block_scales[-attention_dp]
-  - accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_fp8_block_scales[-cuda_graph]
-  - accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_fp8_block_scales[-overlap_scheduler]
-  - accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_fp8_block_scales[-attention_dp-cuda_graph-overlap_scheduler]
-  - accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_fp8_block_scales[mtp_nextn=2-]
-  - accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_fp8_block_scales[mtp_nextn=2-attention_dp]
-  - accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_fp8_block_scales[mtp_nextn=2-cuda_graph]
-  - accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_fp8_block_scales[mtp_nextn=2-overlap_scheduler]
-  - accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_fp8_block_scales[mtp_nextn=2-attention_dp-cuda_graph-overlap_scheduler]
+  - accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_fp8_block_scales[--]
```
Comment (hlu1) [inline tests/integration/test_lists/test-db/l0_h100.yml:35]: Is there some automatic script to update the tests?

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py [hunk 10]
```diff
@@ -257,6 +261,7 @@ class TestDeepSeekV3Lite(LlmapiAccuracyTestHarness):
     MODEL_PATH = f"{llm_models_root()}/DeepSeek-V3-Lite/bf16"
 
     @pytest.mark.skip_less_device_memory(60000)
+    @parametrize_with_ids("torch_compile", [False, True])
```
Comment (hlu1) [inline tests/integration/defs/accuracy/test_llm_api_pytorch.py:264]: So we're going to enable torch.compile tests in CI from now on?

File: tensorrt_llm/_torch/modules/attention.py [hunk 11]
```diff
@@ -239,6 +242,47 @@ def forward(
...
+    )
+
+    mla_layers = extra_attrs.get("mla_layers", None)
+    assert mla_layers is not None, "MLA layers is not registered"
+    mla_layer_ref = mla_layers.get(layer_idx, None)
+    assert mla_layer_ref is not None, f"Cannot find MLA layer for layer {layer_idx}"
+    mla_layer = mla_layer_ref()
+    assert issubclass(type(mla_layer), MLA) or isinstance(
+        mla_layer,
+        MLA), "MLA layer must be a subclass of MLA or an instance of MLA"
```
Comment (hlu1) [inline tensorrt_llm/_torch/modules/attention.py:264]:
```suggestion
    assert isinstance(
        mla_layer,
        MLA), "MLA layer must be an instance of MLA"
```
See https://stackoverflow.com/questions/8107313/isinstance-and-issubclass-return-conflicting-results

File: tensorrt_llm/_torch/attention_backend/flashinfer.py [hunk 12]
```diff
@@ -163,6 +164,19 @@ def __post_init__(self) -> None:
                                                  device='cuda',
                                                  dtype=torch.int)
 
+    def create_cuda_graph_metadata(self,
```
Comment (hlu1) [inline tensorrt_llm/_torch/attention_backend/flashinfer.py:167]: Is this used somewhere?

**Comment by hlu1** [comment]
Added @QiJune  to review the CI/pytorch changes

File: tests/integration/test_lists/test-db/l0_h100.yml (hunk 9 above)
Comment (liji-nv) [reply tests/integration/test_lists/test-db/l0_h100.yml:35]: No, I have to update it one by one. Previously we allow parameter to be None not shown in test name but that was changed.

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py (hunk 10 above)
Comment (liji-nv) [reply tests/integration/defs/accuracy/test_llm_api_pytorch.py:264]: Yes for all passed tests.

File: tensorrt_llm/_torch/attention_backend/flashinfer.py (hunk 12 above)
Comment (liji-nv) [reply tensorrt_llm/_torch/attention_backend/flashinfer.py:167]: Here we just override the create_cuda_graph_metadata in parent class.

File: tensorrt_llm/_torch/custom_ops/cpp_custom_ops.py (hunk 8 above)
Comment (liji-nv) [reply tensorrt_llm/_torch/custom_ops/cpp_custom_ops.py]: We can't. The calculation is not calculated in int but symInt for fake impl. It cannot share the same impl in C++.

File: tests/integration/test_lists/test-db/l0_b200.yml [hunk 13]
```diff
@@ -15,22 +15,34 @@ l0_b200:
...
-  - accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_bfloat16[mtp_nextn=2-cuda_graph]
-  - accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_bfloat16[mtp_nextn=2-overlap_scheduler]
-  - accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_bfloat16[mtp_nextn=2-attention_dp-cuda_graph-overlap_scheduler]
-  - accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_nvfp4[]
-  - accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_nvfp4[fp8kv]
-  - accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_nvfp4[attention_dp]
-  - accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_nvfp4[cuda_graph]
-  - accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_nvfp4[overlap_scheduler]
-  - accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_nvfp4[fp8kv-attention_dp-cuda_graph-overlap_scheduler]
+  - accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_bfloat16[--]
```
Comment (QiJune) [inline tests/integration/test_lists/test-db/l0_b200.yml:18]: Could you please list the CI time of each PyTorch stage before and after this PR? I am afraid this PR will increase CI time greatly.

Comment (liji-nv) [reply tests/integration/test_lists/test-db/l0_b200.yml:18]: | Job | Time after  [/LLM/main/L0_MergeRequest_PR pipeline #3381](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/3381/) | Time before [SingleGPU](https://prod.blsm.nvidia.com/sw-tensorrt-top-1/blue/organizations/jenkins/LLM%2Fmain%2FL0_Test-x86_64/detail/L0_Test-x86_64/10665/pipeline) [MultiGPU](https://prod.blsm.nvidia.com/sw-tensorrt-top-1/blue/organizations/jenkins/LLM%2Fmain%2FL0_Test-x86_64/detail/L0_Test-x86_64/10647/pipeline) |
|----|---|---| 
|[B200_PCIe-PyTorch-1]  | 1h 11m 4s  |  1h 6m 38s |
|[B200_PCIe-PyTorch-2]  | 49m 40s   |  42m 21s |
|[H100_PCIe-PyTorch-1]  | 1h 6m 17s  |   47m 26s |
|[H100_PCIe-PyTorch-2  | 1h 2m 23s | 1h 22m 57s |
|[H100_PCIe-PyTorch-3] | 52m 19s | 1h 14m 47 |
|[DGX_H100-4_GPUs-PyTorch-DeepSeek-1] | 1h 36m 26s | 1h 39m 45s |

The time after is collected with the run last Friday. The change only includes 5 and 6 new tests for H100 and DGX H100. After we fix DP and MTP we may revisit how to cover things without introducing much time.

Comment (liji-nv) [reply tests/integration/test_lists/test-db/l0_b200.yml:18]: Increase about 2~3 min for each new added test that not get skipped.

File: tensorrt_llm/_torch/speculative/mtp.py [hunk 14]
```diff
@@ -751,8 +750,7 @@ def sample_and_accept_draft_tokens(
                                                   device=logits.device)
                 accepted_tokens, num_accepted_tokens = torch.ops.trtllm.mtp_sampling_and_accepted_draft_tokens_op(
                     logits, spec_metadata.draft_tokens, target_tokens_cache,
-                    accepted_tokens, num_accepted_tokens, mtp_num_modules,
-                    batch_size, num_contexts, logits.shape[-1])
+                    mtp_num_modules, batch_size, num_contexts, logits.shape[-1])
```
Comment (lfr-0531) [inline tensorrt_llm/_torch/speculative/mtp.py:753]: Could you make the same changes to `torch.ops.trtllm.mtp_relaxed_acceptance_op`? And then rm those two tensors in L685-L691.

**Review by yizhang-nv — APPROVED** [review]
